src/pyhf/experimental/modifiers.py

```diff
@@ -0,0 +1,181 @@
+from __future__ import annotations
+
+import logging
+from typing import Any, Callable, Sequence
+
+import pyhf
+from pyhf import events, get_backend
+from pyhf.parameters import ParamViewer
+
+log = logging.getLogger(__name__)
+
+__all__ = ["add_custom_modifier"]
+
+
+def __dir__():
+    return __all__
+
+
+try:
+    import numexpr as ne
+except ModuleNotFoundError:
+    log.error(
+        "\nInstallation of the experimental extra is required to use pyhf.experimental.modifiers"
+        + "\nPlease install with: python -m pip install 'pyhf[experimental]'\n",
+        exc_info=True,
+    )
+    raise
+
+
+class BaseApplier:
+    ...
+
+
+class BaseBuilder:
+    ...
+
+
+def _allocate_new_param(
+    p: dict[str, Sequence[float]]
+) -> dict[str, str | bool | int | Sequence[float]]:
+    return {
+        "paramset_type": "unconstrained",
+        "n_parameters": 1,
+        "is_shared": True,
+        "inits": p["inits"],
+        "bounds": p["bounds"],
+        "is_scalar": True,
+        "fixed": False,
+    }
+
+
+def make_func(expression: str, deps: list[str]) -> Callable[[Sequence[float]], Any]:
+    def func(d: Sequence[float]) -> Any:
+        return ne.evaluate(expression, local_dict=dict(zip(deps, d)))
+
+    return func
+
+
+def make_builder(
+    func_name: str, deps: list[str], new_params: dict[str, dict[str, Sequence[float]]]
+) -> BaseBuilder:
+    class _builder(BaseBuilder):
+        is_shared = False
+
+        def __init__(self, config):
+            self.builder_data = {"funcs": {}}
+            self.config = config
+
+        def collect(self, thismod, nom):
+            maskval = bool(thismod)
+            mask = [maskval] * len(nom)
+            return {"mask": mask}
+
+        def append(self, key, channel, sample, thismod, defined_samp):
+            self.builder_data.setdefault(key, {}).setdefault(sample, {}).setdefault(
+                "data", {"mask": []}
+            )
+            nom = (
+                defined_samp["data"]
+                if defined_samp
+                else [0.0] * self.config.channel_nbins[channel]
+            )
+            moddata = self.collect(thismod, nom)
+            self.builder_data[key][sample]["data"]["mask"] += moddata["mask"]
+            if thismod:
+                if thismod["name"] != func_name:
+                    print(thismod)
```

Comment:
```suggestion
```

Reply:
I've removed the prints in https://github.com/scikit-hep/pyhf/pull/1991/commits/bb302c2423f702a489a42c8311dc412388e42bd9 but if we need them back for debugging quickly just rever it.